broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java

```diff
@@ -0,0 +1,152 @@
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.atomix.raft.RaftServer.Role;
+import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessorBuilder;
+import io.camunda.zeebe.logstreams.log.LogStream;
+import io.camunda.zeebe.util.health.HealthMonitor;
+import io.camunda.zeebe.util.sched.future.TestActorFuture;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+class StreamProcessorTransitionStepTest {
+
+  TestPartitionTransitionContext transitionContext = new TestPartitionTransitionContext();
+  final StreamProcessorBuilder streamProcessorBuilder = spy(StreamProcessorBuilder.class);
+  final StreamProcessor streamProcessor = mock(StreamProcessor.class);
+
+  private StreamProcessorTransitionStep step;
+
+  @BeforeEach
+  void setup() {
+    transitionContext.setLogStream(mock(LogStream.class));
+    transitionContext.setComponentHealthMonitor(mock(HealthMonitor.class));
+
+    doReturn(streamProcessor).when(streamProcessorBuilder).build();
+
+    when(streamProcessor.openAsync(anyBoolean())).thenReturn(TestActorFuture.completedFuture(null));
+    when(streamProcessor.closeAsync()).thenReturn(TestActorFuture.completedFuture(null));
+
+    step = new StreamProcessorTransitionStep(streamProcessorBuilder);
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldReInstallStreamProcessor")
+  void shouldCloseExistingStreamProcessor(final Role currentRole, final Role targetRole) {
+    // given
+    if (currentRole != null) {
+      transitionTo(currentRole);
+    }
+
+    // when
+    step.prepareTransition(transitionContext, 1, targetRole).join();
+
+    // then
+    assertThat(transitionContext.getStreamProcessor()).isNull();
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldReInstallStreamProcessor")
+  void shouldReInstallStreamProcessor(final Role currentRole, final Role targetRole) {
+    // given
+    if (currentRole != null) {
+      transitionTo(currentRole);
+    }
+    Mockito.clearInvocations(streamProcessor);
+
+    // when
+    transitionTo(targetRole);
+
+    // then
+    assertThat(transitionContext.getStreamProcessor()).isNotNull();
+    verify(streamProcessor).openAsync(anyBoolean());
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldDoNothing")
+  void shouldNotCloseExistingStreamProcessor(final Role currentRole, final Role targetRole) {
+    // given
+    transitionTo(currentRole);
+
```

Comment:
:thought_balloon: One thing I do not understand is when `clearInvocations(streamProcessor)` is needed and when not.

I was wandering why it's not in setup, but maybe this doesn't work with parameterized tests. Is it not needed here, because it is never invoked in one invocation of the parameterized test?

Or could it maybe be needed in the future, when there are other test methods which have interactions. And then this test could fail, because it doesn't clear the invocations.

Maybe it would be easier to understand if `clearInvocations(streamProcessor)` was at the end of a then block which did verify invocations. Then this would better fit the pattern that the test cleans up after itself.

Just a thought, though. Would love to hear your opinions on this and why you did it this way.

Reply:
In test like the following
```
   // given
    if (currentRole != null) {
      transitionTo(currentRole);
    }
    Mockito.clearInvocations(streamProcessor);

    // when
    transitionTo(targetRole);

    // then  
    assertThat(transitionContext.getStreamProcessor()).isNotNull();
    verify(streamProcessor).openAsync(anyBoolean());
```
The `transitionTo` in "given" interacts with mock. `clearInvocations` is to clear those interactions so that we can then test the expected in "then". Alternatively, what I did for the tests for the rest of the steps which is not yet in this PR is as follows:
```
   // given
    context.setCurrentRole(currentRole)
    if (currentRole != null) {
      context.setStreamProcessor(mockStreamProcessorFromPrevRole);
    }

    // when
    transitionTo(targetRole);

    // then  
    assertThat(transitionContext.getStreamProcessor()).isNotNull().isNotEqualTo(mockStreamProcessorFromPrevRole);
    verify(streamProcessor).openAsync(anyBoolean());
```

In this way there is no real steps called in "given" and thus no invocations to the mock. If this is more clear, I will change the tests to follow this pattern. Wdyt?

Reply:
Love it!